includes/compat/trait-wc-stripe-subscriptions.php

```diff
@@ -244,6 +246,103 @@ public function process_change_subscription_payment_method( $order_id ) {
 		}
 	}
 
+	/**
+	 * Process the payment method change with deferred intent.
+	 *
+	 * @param int $subscription_id
+	 *
+	 * @return array
+	 */
+	public function process_change_subscription_payment_with_deferred_intent( $subscription_id ) {
+		$subscription = wcs_get_subscription( $subscription_id );
+
+		if ( ! $subscription ) {
+			return [
+				'result'   => 'failure',
+				'redirect' => '',
+			];
+		}
+
+		try {
+			$payment_information = $this->prepare_payment_information_from_request( $subscription );
+
+			$this->validate_selected_payment_method_type( $payment_information, $subscription->get_billing_country() );
+
+			$payment_method_id     = $payment_information['payment_method'];
+			$selected_payment_type = $payment_information['selected_payment_type'];
+			$upe_payment_method    = $this->payment_methods[ $selected_payment_type ] ?? null;
+
+			// Retrieve the payment method object from Stripe.
+			$payment_method = $this->stripe_request( 'payment_methods/' . $payment_method_id );
+
+			// Throw an exception when the payment method is a prepaid card and it's disallowed.
+			$this->maybe_disallow_prepaid_card( $payment_method );
+
+			// Create a setup intent, or update an existing one associated with the order.
+			$payment_intent = $this->process_setup_intent_for_order( $subscription, $payment_information );
+
+			// Handle saving the payment method in the store.
+			if ( $payment_information['save_payment_method_to_store'] && $upe_payment_method && $upe_payment_method->get_id() === $upe_payment_method->get_retrievable_type() ) {
+				$this->handle_saving_payment_method(
+					$subscription,
+					$payment_information['payment_method_details'],
+					$selected_payment_type
+				);
+			}
+
+			$redirect           = $this->get_return_url( $subscription );
+			$new_payment_method = $this->get_upe_gateway_id_for_order( $upe_payment_method );
+
+			// If the payment intent requires confirmation or action, redirect the customer to confirm the intent.
+			if ( in_array( $payment_intent->status, [ 'requires_confirmation', 'requires_action' ], true ) ) {
+				// Because we're filtering woocommerce_subscriptions_update_payment_via_pay_shortcode, we need to manually set this delayed update all flag here.
+				if ( isset( $_POST['update_all_subscriptions_payment_method'] ) && wc_clean( wp_unslash( $_POST['update_all_subscriptions_payment_method'] ) ) ) {
+					$subscription->update_meta_data( '_delayed_update_payment_method_all', $new_payment_method );
```

Comment:
Manually setting this same subscription meta as [done here](https://github.com/Automattic/woocommerce-subscriptions-core/blob/6bfbf519df799dde8cd9f39e6c93e820e5170352/includes/class-wc-subscriptions-change-payment-gateway.php#L370-L371) is a bit unfortunate but because we're filtering `woocommerce_subscriptions_update_payment_via_pay_shortcode` to return false, the logic that adds this meta is never ran (notice how [`can_update_all_subscription_payment_methods()`](https://github.com/Automattic/woocommerce-subscriptions-core/blob/6bfbf519df799dde8cd9f39e6c93e820e5170352/includes/class-wc-subscriptions-change-payment-gateway.php#L464-L466) also checks for this filter) so I had to add the meta manually.

Reply:
For additional context, using the same meta lets us call functions like `WC_Subscriptions_Change_Payment_Gateway::will_subscription_update_all_payment_methods()` inside the intent controller class when we're handling changing to 3DS cards but also, when we call `WC_Subscriptions_Change_Payment_Gateway::update_all_payment_methods_from_subscription(), it had built-in code to remove this meta which is handy.